spec/models/user_spec.rb

```diff
@@ -0,0 +1,421 @@
+        digest_token = 'digest_token'
+        digest = double('digest')
+        allow(user).to receive(:send).with(attr_digest).and_return(digest_token)
+        allow(BCrypt::Password).to receive(:new).with(digest_token)
+        .and_return(digest)
+        allow(digest).to receive(:is_password?).with(token).and_return(true)
+        auth_token = user.authentic_token?(attribute, token)
+        expect(auth_token).to be_truthy
+      end
+    end
+
+    describe '#authenticate!'  do
+      let(:attribute) { :attribute }
+
+      context "when password passed is user's password"  do
+        before(:each) do
+          allow(user).to receive(:authenticate).with(user.password)
+            .and_return(user)
+        end
+
+        it 'returns the user'  do
+          expect(user.authenticate!(user.password, attribute)).to eq(user)
+        end
+
+        it 'should not add error to model erros'  do
+          user.authenticate!(user.password, attribute)
+          expect(user.errors).to be_empty
+        end
+      end
+
+      context 'with incorrect password'  do
+        let(:wrong_password) { 'wrong password' }
+        let(:message) { 'message' }
+
+        before(:each) do
+          allow(user).to receive(:authenticate).with(wrong_password)
+            .and_return(false)
+        end
+
+        it "returns false if password passed is user's password"  do
+          expect(user.authenticate!(wrong_password, attribute)).to be_falsy
+        end
+
+        it "adds error to model erros"  do
+          errors = spy('erros')
+          allow(errors).to receive(:add)
+          allow(user).to receive(:errors).and_return(errors)
+          user.authenticate!(wrong_password, attribute, message)
+          expect(errors).to have_received(:add).with(attribute, message)
+        end
+      end
+    end
+
+    describe '#create_reset_digest'  do
+      let(:token) { 'token' }
+      let(:digest) { 'digest' }
+      let(:time_zone) { Time.zone }
+      let(:now) { time_zone.now }
+
+      before(:each) do
+        allow(User).to receive(:new_token).and_return(token)
+        allow(User).to receive(:digest).with(token).and_return(digest)
+        allow(time_zone).to receive(:now).and_return(now)
+        allow(Time).to receive(:zone).and_return(time_zone)
+      end
+
+      it 'updates reset_digest column with new digest token'  do
+        user.create_reset_digest
+        expect(user.reset_digest).to eq(digest)
+      end
+
+      it 'updates reset_sent_at column with new time'  do
+        user.create_reset_digest
+        expect(user.reset_sent_at).to eq(now)
+      end
+
+      it 'calls update_columns with activated true and current time'  do
+        allow(user).to receive(:update)
+        user.create_reset_digest
+        expect(user).to have_received(:update)
+          .with(reset_digest: digest, reset_sent_at: now)
+      end
+    end
+
+    describe '#cropping?'  do
```

Comment:
I noticed that both User and Image models have a cropping method. Could the User_model use the Image cropping method instead of having it's own and we avoid duplicated code?